rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,110 @@
+  * calculated based on the hashes of all the inputs, including build tools
+* `$cas`: output hash, the total hash of all the files under $out, with the derivation name appended, e.g., `qqzyb6bsr5yhh2r5624x01cy87xzn7aa-unit-script-1.12`
+
+## Concept
+
+The basic concept is aliasing equivalent input derivations in such a way that dependent derivations won't need to change if only `$out` changes but not the input derivation contents.
+
+After building a derivation, `$cas` is calculated, and `$out` is renamed to `$cas`. Then, if another build requires the input `$out`, it gets `$cas` instead, and all references to that build input will be `$cas` instead of `$out`. That dependent derivation will also have its input hash calculated with the `$cas` instead of the `$out`.
+
+This means that if 2 different derivations of the same input have a different `$out` but the same `$cas`, any dependent builds will not need to rebuild due to the inputs being different. For example, the 12MB input `poppler-data` is often the same across multiple different input derivations, so many `$out`s for `poppler-data` all result in the same `$cas`. Similarly, a compiler flag change might leave most derivations unchanged.
+
+In order to know which `$out`s refer to a particular `$cas`, symlinks can be used (`$out` pointing to `$cas`), or that data can be stored in the store database. The database can help with doing reverse lookups from `$cas` to all the `$out`s.
+
+## Calculating `$cas`
+
+There is one important corner case that needs special handling: if a derivation refers to itself, it will be referring to `$out`, because `$cas` is not known at the time of the build. This means that each `$out` of a furthermore equivalent build would have a different hash, due to the different `$out`s.
+
+To fix this, the `$cas` calculation has to replace all occurrences of `$out` with an equal-length string of (for example) NULL bytes. After that, `$out` is renamed to `$cas` and all occurrences of `$out` are replaced with `$cas`.
+
+This also means that `$out` and `$cas` should have the same length. The easiest way to achieve that is to use the same hash function for the output hash as used for the input hashes.
+
+To calculate `$cas` we need to include all the data that uniquely defines a derivation: the file contents, case-sensitive names, and the permission bits, traversed in a fixed order, no matter what the filesystem or platform. Not to be included are the owning `uid` of the store and timestamps.
+
+## Distributing derivations
+
+Since `$cas` is only known when `$out` is built, binary caches would need to retain that information. When you look up `$out` to see if it was built already, the response should be _"Yes, this is available as `$cas`"_.
+
+## Maintaining the Nix store
+
+When garbage collecting, the Nix store should also remove `$out` references (be they symlinks or db entries) when removing a `$cas`.
+
+## Micro-optimizations not worth considering
```

Comment:
I would like to eventually moved to an addressing schema used by other systems, e.g. IPFS, which does mean stripping the name and version. But that is follow-up work so your plan is good as first step, unless we are wary about changing hashes/protocols twice (I hope we aren't!).

Reply:
Yeah I do kind of want to so these micro optimizations now or soon.

IIRC, any fixed output with the same hash as an already built one will be copied from the other, but downstream builds see the output hash and thus would need to be built twice. Even with this proposal, the cas still has the name so this caching inconsistency remains.

But if we remove the name (and version) from the CAS, then the rules for substitutions and downstream rebuilds are the same. Any two fixed output derivation that can be substituted for one another at build time are also observationally equivalent downstream. In other words we have a coherent equivalence class (on output hash) that we use everywhere.

Implementatiom wise, we also get to unify things. There is global hash<->build relation, along with an eventually per-user/per-project drv<->hash derivation. For normal derivations, one does the build, and associates the build with the hash (throwing away the build if it is already mapped), and then the hash with the drv. For fixed output derivations, one first associates the drv with the hash, and then only builds the derivation if the hash is not present. So the same two data structures support all caching and substitution.

Reply:
I think you are forgetting the `system` input string field of `derivation` here, which should ultimately also be removed if it is not necessary to describe a difference in how the output is generated.

Reply:
I was thinking about things in the name/`$cas`, not everything in the derivation. Nothing but the name and contents in the proposal influence the `$cas`, right?